lib/Frontend/ParseableInterfaceModuleLoader.cpp

```diff
@@ -0,0 +1,1022 @@
+    subInvocation.getLangOptions().EnableObjCAttrRequiresFoundation = false;
+
+    // Tell the subinvocation to serialize dependency hashes if asked to do so.
+    auto &frontendOpts = subInvocation.getFrontendOptions();
+    frontendOpts.SerializeParseableModuleInterfaceDependencyHashes =
+      serializeDependencyHashes;
+  }
+
+  bool extractSwiftInterfaceVersionAndArgs(
+    swift::version::Version &Vers, llvm::StringSaver &SubArgSaver,
+    SmallVectorImpl<const char *> &SubArgs) {
+    auto FileOrError = swift::vfs::getFileOrSTDIN(fs, interfacePath);
+    if (!FileOrError) {
+      diags.diagnose(diagnosticLoc, diag::error_open_input_file,
+                     interfacePath, FileOrError.getError().message());
+      return true;
+    }
+    auto SB = FileOrError.get()->getBuffer();
+    auto VersRe = getSwiftInterfaceFormatVersionRegex();
+    auto FlagRe = getSwiftInterfaceModuleFlagsRegex();
+    SmallVector<StringRef, 1> VersMatches, FlagMatches;
+    if (!VersRe.match(SB, &VersMatches)) {
+      diags.diagnose(diagnosticLoc,
+                     diag::error_extracting_version_from_parseable_interface);
+      return true;
+    }
+    if (!FlagRe.match(SB, &FlagMatches)) {
+      diags.diagnose(diagnosticLoc,
+                     diag::error_extracting_flags_from_parseable_interface);
+      return true;
+    }
+    assert(VersMatches.size() == 2);
+    assert(FlagMatches.size() == 2);
+    Vers = swift::version::Version(VersMatches[1], SourceLoc(), &diags);
+    llvm::cl::TokenizeGNUCommandLine(FlagMatches[1], SubArgSaver, SubArgs);
+    return false;
+  }
+
+  /// Populate the provided \p Deps with \c FileDependency entries including:
+  ///
+  ///    - All the dependencies mentioned by \p SubInstance's DependencyTracker,
+  ///      that were read while compiling the module.
+  ///
+  ///    - For any file in the latter set that is itself a .swiftmodule
+  ///      living in \p cachePath, all of _its_ dependencies, copied
+  ///      out to avoid having to do recursive scanning when rechecking this
+  ///      dependency in the future.
+  bool collectDepsForSerialization(CompilerInstance &SubInstance,
+                                   SmallVectorImpl<FileDependency> &Deps,
+                                   bool IsHashBased) {
+    auto DTDeps = SubInstance.getDependencyTracker()->getDependencies();
+    SmallVector<StringRef, 16> InitialDepNames(DTDeps.begin(), DTDeps.end());
+    InitialDepNames.push_back(interfacePath);
+    llvm::StringSet<> AllDepNames;
+    for (auto const &DepName : InitialDepNames) {
+      if (AllDepNames.insert(DepName).second && dependencyTracker) {
+        dependencyTracker->addDependency(DepName, /*isSystem*/false);
+      }
+
+      auto DepBuf = getBufferOfDependency(fs, DepName, interfacePath,
+                                          diags, diagnosticLoc);
+      if (!DepBuf)
+        return true;
+      auto Status = getStatusOfDependency(fs, DepName, interfacePath,
```

Comment:
How do we ensure that `DepBuf` and `Status` correspond to the content that was used to compile the module?  Is there some external invariant here?  This seems like something we need to track at the point the file is opened for reading during the compilation.

Reply:
Are you asking what we do if the files have changed during compilation? If that's the case, I don't know if there's anything we _can_ do.

Reply:
You can fstat the file after opening it during the compilation, which guarantees you have a size/timestamp that correspond to the file contents that you read.  The current setup not only might be wrong, but it's undetectably wrong - you could store a newer timestamp/size than what you actually read, and then anyone reading it may incorrectly consider the module up to date.

Reply:
I'm going to address the correctness issue in https://github.com/apple/swift/pull/23084, but I'd like to land this before then.